tests/unit/models/gpflow/test_models.py

```diff
@@ -403,6 +403,7 @@ def test_gaussian_process_regression_trajectory_sampler_has_correct_samples(
         gpr_model(x, _3x_plus_gaussian_noise(x)), use_decoupled_sampler=use_decoupled_sampler
     )
     model.model.likelihood.variance.assign(1e-3)
+    model._posterior.update_cache()
```

Comment:
Its a little worrying that we have to do this here!

Reply:
Perhaps it would be better to not calculate the posterior object until we make the first prediction?  Then we can do whatever model init stuff we want

Reply:
Good suggestion. Done.

Reply:
Ok, it turns out that tensorflow compilation stops us from doing this. I'll have a think if there's another way.

Reply:
For now I've made `update_posterior_cache` a method and clarified the docstring slightly. Not sure if there's much else we can do, as the object need to exist when we compile the wrapper after initialisation. We could decide to always regenerate the cache the first time we call predict, but that feels wasteful. @jesnie : any ideas?

Reply:
As an aside, it's worth remembering that you only need to update the cache if you modified the model after creating the model wrapper. You can play with model init stuff all you want as long as you wrap it afterwards.

Reply:
> For now I've made `update_posterior_cache` a method and clarified the docstring slightly. Not sure if there's much else we can do, as the object need to exist when we compile the wrapper after initialisation. We could decide to always regenerate the cache the first time we call predict, but that feels wasteful. @jesnie : any ideas?

Not really. I have been thinking the last couple of weeks, that it would be cool with some kind of framework that dynamically decide what to recompute when variables change, but I don't think there really is a good solution. For the foreseeable future I think we have to manually update stuff whenever a variable is changed.

Reply:
So whats the situation with this change?